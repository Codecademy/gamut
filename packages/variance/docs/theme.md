# Theme Construction

In order to use variance with your custom theme you must redeclare the emotion global `Theme`.

```tsx
import { theme } from './theme';

export type ThemeShape = typeof theme;

declare module '@emotion/react' {
  export interface Theme extends ThemeShape {}
}
```

To help you create this strongly typed theme we've provided a few helpers to make it easier to progressively create a strongly typed and dynamic theme that will adhere to the correct types.

## `createTheme()`

This method is a set of chainable methods that will progressively add tokens, relationships, and core features to a theme object.

While many frameworks restrict theme scales to very specific sets we leave many of the particulars up to the user to define as the only key that we require for baseline functionality is `breakpoints`.

We do however offer optional but opinionated handling of both `colors` + `colorModes`.

### Usage

```tsx
import { ThemeProvider, Global, css } from '@emotion/react';

export const {
  /**
   * A variance compatible theme with strong typings.
   * Values on theme may include references to more complex CSS variables
   * without using the variables through a provider.
   */
  theme,
  /**
   * Serialized CSS variables that correspond with theme references.
   * These are scoped to 2 main areas currently root and colorMode
   */
  variables: {
    /**
     *  Top level variables to available for all HTML
     */
    root,
    /** This is still placed on the root but is kept separate from the
     * general variables as they are context dependent.
     */
    colorMode,
  },
  /**
   * Method that returns the static value for a named color without having to check the computed value or importing tokens and possible nested references.
   */
  getColorValue,
} = createTheme({
  breakpoints: {
    xs: '@media screen and (min-width: 480px)',
    sm: '@media screen and (min-width: 768px)',
    md: '@media screen and (min-width: 1024px)',
    lg: '@media screen and (min-width: 1200px)',
    xl: '@media screen and (min-width: 1440px)',
  },
  spacing: {
    4: '0.25rem',
    8: '0.5rem',
    12: '0.75rem',
    16: '1rem',
  },
})
  .addColors({
    white: '#ffffff',
    hyper: '3A10E5',
    navy: '#10162f',
    yellow: '#FFD300',
  })
  .createColorModes('light', {
    light: {
      primary: 'hyper',
      secondary: 'navy',
      text: 'navy',
      background: 'white',
    },
    dark: {
      primary: 'yellow',
      secondary: 'white',
      text: 'white',
      background: 'navy',
    },
  })
  .addVariables('spacing')
  .build();

/**
 * To use your new theme you must do 3 things.
 * 1. Declare your theme shape as the emotion theme.
 * 2. Wrap your application in a ThemeProvider
 * 3. And add your variables to the global styles.
 */

export type ThemeShape = typeof theme;

declare module '@emotion/react' {
  export interface Theme extends ThemeShape {}
}

const App = () => {
  return (
    <ThemeProvider theme={theme}>
      <Global styles={css({ ':root': root })} />
      <Global styles={css({ ':root': colorMode })} />
      <Component />
    </ThemeProvider>
  );
};
```

### Methods

A theme creator method that progressively build and decorate a type safe theme object. This allows us to create a build separate themes that will be compatible with variance props. The raw theme must have a valid breakpoint scale to ensure breakpoint variable serialization.

**Standard Methods** - These are the core methods to add scales and variables to the theme object

- `addScale(scaleKey, updateFunction)` - Adds a set of tokens to the theme on the provided scale key. The current theme can be used as reference for this new scale for shared value reference.
- `updateScale(scaleKey, updateFunction)` - Updates an existing scale with new or computed value without them being explicitly declared. This can also be used to update or extend a theme for a different context (restrictions or expansions).
- `createScaleVariables(scaleKey)` - Takes a top level key of the current theme and serialized them as CSS variables. The theme will include references to CSS Variables and adds the actual token values to the root scope of our variable object.

**Special Methods** - these have specific behavior that is non standard, these are both required for all features to work

- `addColors(tokens)` - Adds color tokens to the theme and creates root color variables by default. Calling this method required to access `getColorValue` and `addColorModes`. You will not be able to add colors through `addScale` and be able to build color modes as they have different internal behaviors. However if you do not want either features you may use it without any issue.
- `addColorModes(initialMode, colorModes)` - This method takes a configuration of color aliases that have semantic meaning between contexts such as `light` and `dark` modes. This will take a map of modes and an initial mode.
  - Colors must exist on the theme for this to work, call this method after `addColors` to ensure this works correctly.
  - This creates CSS variables for the initial color mode and adds them to the specific bucket at the root scope and ensures that nested color variable references behave correctly (if variables change these will too).

**Finalization**

- `build()` - Called when all mutations are finished and we get the finalized and fully typed design system objects.

# Other Utilities

These are some under the hood utilities that you can use outside of root theme creation. `createTheme` may use some of these internally but they can also be used independently.

## `serializeTokens()`

### Arguments

- `tokens` any set of tokens.
- `prefix` a string to prefix any tokens
- `theme` to reference existing tokens specifically for breakpoints.

This method predictably maps token literal values to CSS variables. We use this to store relational or contextual information in a single reference.

### Usage

```tsx
const { tokens, variables } = serializeTokens({
  black: '#000000',
  white: '#FFFFFF'
}, 'color', {});

// tokens
{ black: 'var(--color-black)', white: 'var(--color-white)' }

// variables
{ '--color-black': '#000000', '--color-white': '#FFFFFF' };
```

This will also work with possible nested selectors like breakpoints:

```tsx
const {
  tokens, //An object of the same keys as the first argument but with values that point to variable references
  variables // Valid CSS variables that are prefixed with the same keys.
} = serializeTokens({
  height: { _: '4rem', lg: '5rem' },
}, 'header', theme);

// tokens
{ height: 'var(--header-height)' }

// variables
{
  '--header-height': '4rem',
  '@media screen and (min-width: 1024px)': {
    '--header-height': '5rem',
  }
}
```
