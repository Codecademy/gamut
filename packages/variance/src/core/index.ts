export * from './compose';
export * from './createProps';
export * from './createCss';
export * from './createVariants';
