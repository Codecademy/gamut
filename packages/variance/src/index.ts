export { variance } from './core';
export * from './createTheme';
export * from './types/props';
export * from './transforms';
export * from './scales/createScale';
