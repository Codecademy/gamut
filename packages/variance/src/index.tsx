export * from './core';
export * from './types/props';
