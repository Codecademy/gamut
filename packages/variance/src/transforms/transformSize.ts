import { isNumber } from 'lodash';

export const percentageOrAbsolute = (coordinate: number) => {
  if (coordinate === 0) {
    return coordinate;
  }
  if (coordinate <= 1 && coordinate >= -1) {
    return `${coordinate * 100}%`;
  }
  return `${coordinate}px`;
};

const valueWithUnit = /(-?\d*\.?\d+)(%|\w*)/;

export const transformSize = (value: string | number) => {
  if (isNumber(value)) {
    return percentageOrAbsolute(value);
  }

  if (value.includes('calc')) {
    return value;
  }

  const [match, number, unit] = valueWithUnit.exec(value) || [];

  if (match === undefined) {
    return value;
  }

  const numericValue = parseFloat(number);

  return !unit ? percentageOrAbsolute(numericValue) : `${numericValue}${unit}`;
};
