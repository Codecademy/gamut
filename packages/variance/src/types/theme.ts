interface Breakpoints<T = string> {
  xs: T;
  sm: T;
  md: T;
  lg: T;
  xl: T;
}

export interface BaseTheme {
  breakpoints: Breakpoints;
}

declare module '@emotion/react' {
  export interface Theme extends BaseTheme {}
}
