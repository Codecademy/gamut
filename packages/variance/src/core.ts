import { get, identity, merge } from 'lodash';

import {
  AbstractParser,
  AbstractPropTransformer,
  Parser,
  Prop,
  PropTransformer,
  TransformerMap,
} from './types/config';
import { AbstractTheme, CSSObject, ThemeProps } from './types/props';
import { AllUnionKeys, KeyFromUnion } from './types/utils';
import { orderPropNames } from './utils/propNames';
import {
  arrayParser,
  isMediaArray,
  isMediaMap,
  objectParser,
  parseBreakpoints,
} from './utils/responsive';

export const variance = {
  withTheme<T extends AbstractTheme>(theme: T) {
    const breakpoints = parseBreakpoints(theme);

    return {
      // Parser to handle any set of configured props
      createParser<Config extends Record<string, AbstractPropTransformer<T>>>(
        config: Config
      ): Parser<T, Config> {
        const propNames = orderPropNames(config);

        const parser = (props: ThemeProps<T>) => {
          const styles = {};
          const { map, array } = breakpoints;

          // Loops over all prop names on the configured config to check for configured styles
          propNames.forEach((prop) => {
            const property = config[prop];
            const value = get(props, prop);

            switch (typeof value) {
              case 'string':
              case 'number':
                return Object.assign(
                  styles,
                  property.styleFn(value, prop, props)
                );
              // handle any props configured with the responsive notation
              case 'object':
                // If it is an array the order of values is smallest to largest: [base, xs, ...]
                if (isMediaArray(value)) {
                  return merge(
                    styles,
                    arrayParser(value, props, property, array)
                  );
                }
                // Check to see if value is an object matching the responsive syntax and generate the styles.
                if (isMediaMap(value)) {
                  return merge(
                    styles,
                    objectParser(value, props, property, map)
                  );
                }
            }
          });
          return styles;
        };
        // return the parser function with the resulting meta information for further composition
        return Object.assign(parser, { propNames, config });
      },
      // Given a single property configuration enrich the config with a transform function
      // that traverses the properties the function is responsible for.
      createTransform<P extends string, Config extends Prop<T>>(
        prop: P,
        config: Config
      ): PropTransformer<T, P, Config> {
        const transform = config.transform ?? identity;
        const properties = config.properties
          ? config.properties
          : [config.property];

        return {
          ...config,
          prop,
          styleFn: (value, prop, props) => {
            const styles: CSSObject = {};
            let scaleVal: string | number;
            switch (typeof config.scale) {
              case 'string': {
                const path = `theme.${config.scale}.${value}`;
                scaleVal = get(props, path, value);
                break;
              }
              case 'object': {
                scaleVal = get(config.scale, `${value}`);
                break;
              }
              default:
                scaleVal = value as string | number;
            }

            // for each property look up the scale value from theme if passed and apply any
            // final transforms to the value
            properties.forEach((property) => {
              styles[property] = transform(scaleVal, property, props);
            });
            // return the resulting styles object
            return styles;
          },
        };
      },
      compose<Args extends AbstractParser<T>[]>(...parsers: Args) {
        type MergedParser = {
          [K in AllUnionKeys<Args[number]['config']>]: KeyFromUnion<
            Args[number]['config'],
            K
          >;
        };
        return this.createParser(
          parsers.reduce(
            (carry, parser) => ({ ...carry, ...parser.config }),
            {}
          ) as MergedParser
        );
      },
      create<Config extends Record<string, Prop<T>>>(config: Config) {
        const transforms = {} as TransformerMap<T, Config>;

        // Create a transform function for each of the props
        for (const prop in config) {
          if (typeof prop === 'string') {
            transforms[prop] = this.createTransform(prop, config[prop]);
          }
        }

        // Create a parser that handles all the props within the config
        return this.createParser(transforms);
      },
    };
  },
};
