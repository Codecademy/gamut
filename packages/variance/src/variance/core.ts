import { get, identity, isObject, merge } from 'lodash';

import {
  AbstractParser,
  AbstractPropTransformer,
  CSS,
  Parser,
  Prop,
  PropTransformer,
  TransformerMap,
  Variant,
} from './types/config';
import {
  AbstractTheme,
  BreakpointCache,
  CSSObject,
  ThemeProps,
} from './types/props';
import { AllUnionKeys, KeyFromUnion } from './types/utils';
import { isMediaArray, isMediaMap } from './utils';

export const variance = {
  withTheme: function <T extends AbstractTheme>() {
    return {
      getBreakpoints: function (props: { theme: T }): BreakpointCache<T> {
        const breakpoints = props?.theme?.breakpoints;
        if (!breakpoints) return null;

        const { xs, sm, md, lg, xl } = breakpoints;
        return {
          map: breakpoints,
          array: [xs, sm, md, lg, xl],
        };
      },
      createParser: function <
        Config extends Record<string, AbstractPropTransformer<T>>
      >(config: Config): Parser<T, Config> {
        let cache: BreakpointCache<T>;
        const propNames = Object.keys(config);

        //
        const parser = (props: { theme: T }) => {
          const styles = {};
          const breakpoints =
            cache === undefined ? this.getBreakpoints(props) : cache;
          if (cache === undefined) {
            cache = breakpoints;
          }
          propNames.forEach((prop) => {
            const transform = config[prop].styleFn;
            const value = get(props, prop);
            switch (typeof value) {
              case 'string':
              case 'number':
                Object.assign(styles, transform(value, prop, props));
                break;
              case 'object':
                if (!breakpoints) return;
                const breakpointStyles = {};

                if (isMediaArray(value)) {
                  const [base, ...rest] = value;

                  if (base) Object.assign(styles, transform(base, prop, props));

                  rest.forEach((val, i) => {
                    const breakpointKey = breakpoints.array[i];

                    merge(breakpointStyles, {
                      [breakpointKey]: transform(val, prop, props),
                    });
                  });
                }

                if (isMediaMap(value)) {
                  const { base = undefined, ...rest } = value;

                  if (base) Object.assign(styles, transform(base, prop, props));

                  Object.keys(rest).forEach((bp: keyof typeof rest) => {
                    const breakpointKey = get(breakpoints, `map.${bp}`);
                    merge(breakpointStyles, {
                      [breakpointKey]: transform(rest[bp], prop, props),
                    });
                  });
                }
                merge(styles, breakpointStyles);
            }
          });
          return styles;
        };

        return Object.assign(parser, { propNames, config });
      },
      createTransform: function <P extends string, Config extends Prop<T>>(
        prop: P,
        config: Config
      ): PropTransformer<T, P, Config> {
        const transform = config.transform ?? identity;
        const properties = config.properties
          ? config.properties
          : [config.property];

        return {
          ...config,
          prop,
          styleFn: (value, prop, props) => {
            const styles: CSSObject = {};
            properties.forEach((property) => {
              styles[property] = transform(
                get(props, `theme.${config.scale}.${value}`, value),
                prop,
                props
              );
            });
            return styles;
          },
        };
      },
      compose: function <Args extends AbstractParser<T>[]>(...parsers: Args) {
        type MergedParser = {
          [K in AllUnionKeys<Args[number]['config']>]: KeyFromUnion<
            Args[number]['config'],
            K
          >;
        };
        return this.createParser(
          parsers.reduce(
            (carry, parser) => ({ ...carry, ...parser.config }),
            {}
          ) as MergedParser
        );
      },
      createCss: function <
        Config extends Record<string, Prop<T>>,
        P extends Parser<T, TransformerMap<T, Config>>
      >(config: Config): CSS<T, P> {
        const parser = this.create(config);

        return (cssProps) => {
          let cache: CSSObject;
          const selectors = Object.keys(cssProps).filter((key) =>
            key.match(/(&|>|\+|~)/g)
          );

          return ({ theme }) => {
            if (cache) return cache;
            const css = parser({ ...cssProps, theme });
            selectors.forEach((selector) => {
              const selectorConfig = cssProps[selector];
              if (isObject(selectorConfig)) {
                css[selector] = parser(
                  Object.assign(selectorConfig, { theme })
                );
              }
            });
            cache = css;

            return cache;
          };
        };
      },
      createVariant: function <
        Config extends Record<string, Prop<T>>,
        P extends Parser<T, TransformerMap<T, Config>>
      >(config: Config): Variant<T, P> {
        const css: CSS<T, P> = this.createCss(config);

        return (variants, options) => {
          type Keys = keyof typeof variants;
          const prop = options?.prop || 'variant';
          const defaultVariant = options?.defaultVariant;

          const variantFns = {} as Record<
            Keys,
            (props: ThemeProps<T>) => CSSObject
          >;

          Object.keys(variants).forEach((key) => {
            const variantKey = key as Keys;
            const cssProps = variants[variantKey];
            variantFns[variantKey] = css(cssProps);
          });

          return ({ [prop]: selected = defaultVariant, ...props }) => {
            if (!selected) return {};
            return variantFns[selected](props);
          };
        };
      },
      create: function <Config extends Record<string, Prop<T>>>(
        config: Config
      ) {
        const transforms = {} as TransformerMap<T, Config>;

        // Create a transform function for each of the props
        for (const prop in config) {
          transforms[prop] = this.createTransform(prop, config[prop]);
        }

        // Create a parser that handles are the props
        return this.createParser(transforms);
      },
    };
  },
};
