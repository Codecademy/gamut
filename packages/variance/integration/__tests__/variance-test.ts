import { variance } from '../../src/core';
import { transformSize } from '../../src/transforms/transformSize';
import { theme } from '../__fixtures__/theme';

const space = variance.create({
  margin: { property: 'margin', scale: 'spacing' },
  padding: { property: 'padding', scale: 'spacing' },
});

const layout = variance.create({
  width: {
    property: 'width',
    transform: (val: string) => `${parseInt(val, 10) / 16}rem`,
  },
  height: {
    property: 'height',
    transform: (val: string) => `${parseInt(val, 10) / 16}rem`,
  },
});
const rem = (val: number) => `${val / 16}rem`;

type Assert<X, Y> = X extends Y ? true : false;

describe('style props', () => {
  describe('parsers', () => {
    it('has the correct config', () => {
      const propNamesRestricted: Assert<
        typeof space['propNames'],
        ('margin' | 'padding')[]
      > = true;

      expect(propNamesRestricted);

      expect(space.propNames.sort()).toEqual(['padding', 'margin'].sort());
    });
    it('renders styles', () => {
      expect(space({ margin: 4, theme })).toEqual({ margin: '0.25rem' });
    });
    it('renders media query arrays styles', () => {
      const sizes = [4, 8, 16, 24, 32, 48] as const;
      expect(space({ margin: sizes, theme })).toEqual({
        margin: rem(sizes[0]),
        XS: {
          margin: rem(sizes[1]),
        },
        SM: {
          margin: rem(sizes[2]),
        },
        MD: {
          margin: rem(sizes[3]),
        },
        LG: {
          margin: rem(sizes[4]),
        },
        XL: {
          margin: rem(sizes[5]),
        },
      });
    });
    it('renders media query arrays styles', () => {
      const sizes = [4, 8, 16, 24, 32, 48] as const;

      expect(
        space({
          margin: sizes,
          padding: sizes,
          theme,
        })
      ).toEqual({
        margin: rem(sizes[0]),
        padding: rem(sizes[0]),
        XS: {
          margin: rem(sizes[1]),
          padding: rem(sizes[1]),
        },
        SM: {
          margin: rem(sizes[2]),
          padding: rem(sizes[2]),
        },
        MD: {
          margin: rem(sizes[3]),
          padding: rem(sizes[3]),
        },
        LG: {
          margin: rem(sizes[4]),
          padding: rem(sizes[4]),
        },
        XL: {
          margin: rem(sizes[5]),
          padding: rem(sizes[5]),
        },
      });
    });
    it('renders media map arrays styles', () => {
      const sizes = { _: 4, xs: 8, sm: 16, md: 24, lg: 32, xl: 48 } as const;

      expect(space({ margin: sizes, theme })).toEqual({
        margin: rem(sizes._),
        XS: {
          margin: rem(sizes.xs),
        },
        SM: {
          margin: rem(sizes.sm),
        },
        MD: {
          margin: rem(sizes.md),
        },
        LG: {
          margin: rem(sizes.lg),
        },
        XL: {
          margin: rem(sizes.xl),
        },
      });
    });
    it('renders media map arrays styles', () => {
      const sizes = { _: 4, xs: 8, sm: 16, md: 24, lg: 32, xl: 48 } as const;

      expect(space({ margin: sizes, padding: sizes, theme })).toEqual({
        margin: rem(sizes._),
        padding: rem(sizes._),
        XS: {
          margin: rem(sizes.xs),
          padding: rem(sizes.xs),
        },
        SM: {
          margin: rem(sizes.sm),
          padding: rem(sizes.sm),
        },
        MD: {
          margin: rem(sizes.md),
          padding: rem(sizes.md),
        },
        LG: {
          margin: rem(sizes.lg),
          padding: rem(sizes.lg),
        },
        XL: {
          margin: rem(sizes.xl),
          padding: rem(sizes.xl),
        },
      });
    });
    it('passes through falsy values for basic props', () => {
      expect(space({ margin: 0, theme })).toEqual({ margin: 0 });
    });
    it('passes through falsy values for object syntax props', () => {
      expect(space({ margin: { sm: 16, md: 0 }, theme })).toEqual({
        SM: { margin: '1rem' },
        MD: { margin: 0 },
      });
    });
    it('passes through falsy values for array syntax props', () => {
      expect(space({ margin: [, , 16, 0], theme })).toEqual({
        SM: { margin: '1rem' },
        MD: { margin: 0 },
      });
    });
    it('transforms props', () => {
      const res = { height: '1.5rem' };
      expect(layout({ height: '24px', theme })).toEqual(res);
    });
    it('transforms scalar values only if scale is present', () => {
      const doubleSpace = variance.create({
        p: {
          property: 'padding',
          scale: 'spacing',
          transform: (val: number) => `calc(${val} * 2)`,
        },
      });

      expect(doubleSpace({ theme, p: 16 })).toEqual({
        padding: 'calc(1rem * 2)',
      });

      expect(doubleSpace({ theme, p: 'initial' })).toEqual({
        padding: 'initial',
      });
    });
  });
  describe('compose', () => {
    it('combines multiple parsers into one parser', () => {
      const composed = variance.compose(layout, space);

      expect(
        composed({
          height: '24px',
          padding: [4, 16],
          theme,
        })
      ).toEqual({
        height: '1.5rem',
        padding: '0.25rem',
        XS: {
          padding: '1rem',
        },
      });
    });
  });
});

describe('css', () => {
  const marginTransform = jest.fn();

  const css = variance.createCss({
    width: { property: 'width', transform: transformSize },
    height: { property: 'height', transform: transformSize },
    margin: {
      property: 'margin',
      scale: theme.spacing,
      transform: marginTransform,
    },
    padding: { property: 'padding', scale: theme.spacing },
    boxShadow: {
      property: 'boxShadow',
    },
  });

  beforeEach(() => {
    jest.resetAllMocks();
    marginTransform.mockImplementation((val) => val);
  });

  it('creates a CSS Function', () => {
    expect(css).toBeDefined();
  });
  it('produces css', () => {
    const returnedFn = css({ margin: 4, width: '100%', height: '50' });

    expect(returnedFn({ theme })).toEqual({
      margin: '0.25rem',
      width: '100%',
      height: '50px',
    });
  });
  it('works with media queries', () => {
    const returnedFn = css({
      width: ['100%', '200%'],
      height: '50',
    });

    expect(returnedFn({ theme })).toEqual({
      width: '100%',
      XS: { width: '200%' },
      height: '50px',
    });
  });
  it('allows selectors', () => {
    const returnedFn = css({
      width: ['100%', '200%'],
      '&:hover': {
        width: '50%',
      },
    });

    expect(returnedFn({ theme })).toEqual({
      width: '100%',
      XS: { width: '200%' },
      '&:hover': {
        width: '50%',
      },
    });
  });
  it('allows selectors with media queries', () => {
    const returnedFn = css({
      width: ['100%', '200%'],
      boxShadow: [({ colors }) => `0px 0px 0px 0px ${colors.black}`],
      '&:hover': {
        width: ['50%', '25%'],
      },
    });

    expect(returnedFn({ theme })).toEqual({
      width: '100%',
      boxShadow: '0px 0px 0px 0px var(--color-black)',
      XS: { width: '200%' },
      '&:hover': {
        width: '50%',
        XS: { width: '25%' },
      },
    });
  });

  it('allows static valid CSS to pass through', () => {
    const returnedFn = css({
      display: 'grid',
      width: ['100%', '200%'],
    });

    expect(returnedFn({ theme })).toEqual({
      display: 'grid',
      width: '100%',
      XS: { width: '200%' },
    });
  });

  it('allows static valid CSS to pass through within selectors', () => {
    const returnedFn = css({
      display: 'grid',
      '&:hover': {
        display: 'none',
      },
    });

    expect(returnedFn({ theme })).toEqual({
      display: 'grid',
      '&:hover': { display: 'none' },
    });
  });

  it('caches the response', () => {
    const returnedFn = css({
      margin: 4,
    });

    expect(marginTransform).toHaveBeenCalledTimes(0);

    returnedFn({ theme });

    expect(marginTransform).toHaveBeenCalledTimes(1);

    returnedFn({ theme });

    expect(marginTransform).toHaveBeenCalledTimes(1);
  });
});

describe('variants', () => {
  const marginTransform = jest.fn();

  const variant = variance.createVariant({
    width: { property: 'width', transform: transformSize },
    height: { property: 'height', transform: transformSize },
    margin: {
      property: 'margin',
      scale: 'spacing',
      transform: marginTransform,
    },
    padding: { property: 'padding', scale: 'spacing' },
  });

  beforeEach(() => {
    jest.resetAllMocks();
    marginTransform.mockImplementation((val) => val);
  });

  it('creates a variant function', () => {
    const myVariant = variant({
      variants: {
        cool: {
          margin: 4,
          width: ['100%', '200%'],
        },
      },
    });

    expect(myVariant({ theme, variant: 'cool' })).toEqual({
      width: '100%',
      margin: '0.25rem',
      XS: { width: '200%' },
    });
  });
  it('has a default variant', () => {
    const myVariant = variant({
      defaultVariant: 'cool',
      variants: {
        cool: {
          width: ['100%', '200%'],
        },
        beans: {
          height: 16,
        },
      },
    });

    expect(myVariant({ theme })).toEqual({
      width: '100%',
      XS: { width: '200%' },
    });

    expect(myVariant({ theme, variant: 'beans' })).toEqual({ height: '16px' });
  });
  it('has a customized key', () => {
    const myVariant = variant({
      prop: 'sweet',
      variants: {
        cool: {
          width: '100%',
        },
      },
    });

    expect(myVariant({ theme, sweet: 'cool' })).toEqual({
      width: '100%',
    });
  });
  it('allows variant props with selectors', () => {
    const myVariant = variant({
      prop: 'sweet',
      variants: {
        cool: {
          width: '100%',
          '&:hover': {
            width: '50%',
          },
        },
      },
    });

    expect(myVariant({ theme, sweet: 'cool' })).toEqual({
      width: '100%',
      '&:hover': {
        width: '50%',
      },
    });
  });

  it('allows variant props with selectors with media queries', () => {
    const myVariant = variant({
      prop: 'sweet',
      variants: {
        cool: {
          width: '100%',
          '&:hover': {
            width: ['50%', '25%'],
          },
        },
        story: {
          margin: 0,
        },
      },
    });

    expect(myVariant({ theme, sweet: 'cool' })).toEqual({
      width: '100%',
      '&:hover': {
        width: '50%',
        XS: {
          width: '25%',
        },
      },
    });
  });
  it('caches the variant once called', () => {
    const myVariant = variant({
      variants: {
        cool: {
          margin: 4,
        },
      },
    });
    myVariant({ theme });

    expect(marginTransform).toHaveBeenCalledTimes(0);

    myVariant({ theme, variant: 'cool' });

    expect(marginTransform).toHaveBeenCalledTimes(1);

    myVariant({ theme, variant: 'cool' });

    expect(marginTransform).toHaveBeenCalledTimes(1);
  });
  it('caches each variant individually', () => {
    const myVariant = variant({
      variants: {
        cool: {
          margin: 4,
        },
        beans: {
          margin: 8,
        },
        world: {
          margin: 16,
        },
      },
    });

    expect(marginTransform).toHaveBeenCalledTimes(0);

    myVariant({ theme, variant: 'cool' });

    expect(marginTransform).toHaveBeenCalledTimes(1);

    myVariant({ theme, variant: 'cool' });
    myVariant({ theme, variant: 'beans' });

    expect(marginTransform).toHaveBeenCalledTimes(2);

    myVariant({ theme, variant: 'cool' });
    myVariant({ theme, variant: 'beans' });
    myVariant({ theme, variant: 'world' });

    expect(marginTransform).toHaveBeenCalledTimes(3);

    myVariant({ theme, variant: 'cool' });
    myVariant({ theme, variant: 'beans' });
    myVariant({ theme, variant: 'world' });

    expect(marginTransform).toHaveBeenCalledTimes(3);
  });
  it('takes base style css object', () => {
    const baseVariants = variant({
      base: {
        margin: 4,
        padding: [, 8],
        '&:hover': {
          padding: [, 32],
        },
      },
      variants: {
        big: {
          padding: 16,
          margin: [, 8],
          '&:hover': {
            margin: [, 16],
          },
        },
        small: {
          padding: [, 16],
        },
      },
    });

    expect(baseVariants({ variant: 'big', theme })).toEqual({
      margin: '0.25rem',
      padding: '1rem',
      XS: {
        padding: '0.5rem',
        margin: '0.5rem',
      },
      '&:hover': {
        XS: {
          padding: '2rem',
          margin: '1rem',
        },
      },
    });

    expect(baseVariants({ variant: 'small', theme })).toEqual({
      margin: '0.25rem',
      XS: {
        padding: '1rem',
      },
      '&:hover': {
        XS: {
          padding: '2rem',
        },
      },
    });
  });
});

describe('states', () => {
  const marginTransform = jest.fn();

  const states = variance.createStates({
    width: { property: 'width', transform: transformSize },
    height: { property: 'height', transform: transformSize },
    margin: {
      property: 'margin',
      scale: 'spacing',
      transform: marginTransform,
    },
    padding: { property: 'padding', scale: 'spacing' },
  });

  beforeEach(() => {
    jest.resetAllMocks();
    marginTransform.mockImplementation((val) => val);
  });

  it('creates a state variant function', () => {
    const myStates = states({
      cool: {
        margin: 4,
        width: ['100%', '200%'],
      },
      beans: {
        border: '1px solid blue',
      },
    });

    expect(myStates({ theme, cool: true })).toEqual({
      width: '100%',
      margin: '0.25rem',
      XS: { width: '200%' },
    });

    expect(myStates({ theme, cool: true, beans: true })).toEqual({
      width: '100%',
      margin: '0.25rem',
      border: '1px solid blue',
      XS: { width: '200%' },
    });
  });
  it('progressively overrides based on the order of the enabled', () => {
    const myStates = states({
      cool: {
        margin: 4,
      },
      beans: {
        margin: 8,
      },
      dude: {
        margin: 16,
      },
    });

    expect(myStates({ theme, cool: true })).toEqual({
      margin: '0.25rem',
    });

    expect(myStates({ theme, cool: true, beans: true })).toEqual({
      margin: '0.5rem',
    });

    expect(myStates({ theme, cool: true, beans: true, dude: true })).toEqual({
      margin: '1rem',
    });
  });
});
