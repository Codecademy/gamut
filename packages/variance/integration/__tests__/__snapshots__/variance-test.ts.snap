// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`style props parsers media query styles Equivalent Syntax SM and MD -  [ <2 empty items>, 16, 0, [length]: 4 ] === { sm: 16, md: 0 } 1`] = `
Object {
  "MD": Object {
    "margin": 0,
  },
  "SM": Object {
    "margin": "1rem",
  },
}
`;

exports[`style props parsers media query styles Equivalent Syntax all media queries -  [ 4, 8, 16, 24, 32, 48, [length]: 6 ] === { _: 4, xs: 8, sm: 16, md: 24, lg: 32, xl: 48 } 1`] = `
Object {
  "LG": Object {
    "margin": "2rem",
  },
  "MD": Object {
    "margin": "1.5rem",
  },
  "SM": Object {
    "margin": "1rem",
  },
  "XL": Object {
    "margin": "3rem",
  },
  "XS": Object {
    "margin": "0.5rem",
  },
  "margin": "0.25rem",
}
`;

exports[`style props parsers media query styles Equivalent Syntax no media query base value -  [ 4, [length]: 1 ] === { _: 4 } 1`] = `
Object {
  "margin": "0.25rem",
}
`;

exports[`style props parsers media query styles Equivalent Syntax only MD with undefined keys -  [ <3 empty items>, 0, [length]: 4 ] === { md: 0, _: undefined, sm: undefined } 1`] = `
Object {
  "MD": Object {
    "margin": 0,
  },
}
`;

exports[`style props parsers media query styles Equivalent Syntax only XS -  [ <1 empty item>, 4, [length]: 2 ] === { xs: 4 } 1`] = `
Object {
  "XS": Object {
    "margin": "0.25rem",
  },
}
`;
